### ISSUE TITLE
Behavior around BigQuery partitioned table

### ISSUE THREAD

**Issue by tmshn** [dbt-labs/dbt#3717]
Replacing partition(s) is a good strategy to partialy update time-growing BigQuery table idempotently.

While it is supported in dbt, its behavior is hard to understand, and actually there seem to be some bugs.

| Partitioning type           | dbt configuration                                                  |
| :-------------------------- | :----------------------------------------------------------------- |
| Ingestion time partitioning | `materialization='table'` and specifies `partitions`               |
| Column based partitioning   | `materialization='incremental', incremental_strategy='insert_overwrite'` |

## Problems

1. Partial update (i.e.: non full replace) of ingestion-time partitioned table is achieved using `materialization='table'` unlike column partitioned table
2. Cannot specify time granularity for ingestion-time partitioned table
3. With `incremental_strategy='insert_overwrite'`, specifying both `partition_by` and `partitions` has two bugs: 1. it raises type error, 2: it trys to insert data out of specified partition
4. With some combination of options, specified value is ignored, but undocumented
    - `materialization='table'` and specify `partitions` → `partition_by` is ignored
    - `incremental_strategy='insert_overwrite'` → `unique_key` is ignored
    - `incremental_strategy='merge'` → `partitions` is ignored

## Suggested solutions

1. Allow `partition_by = {'field': '_PARTITIONTIME'}` to make ingestion time partitioned table
2. Make `partitions` option to work only with `'insert_overwrite'`-ing ingestion time partitioned table
    - This change makes `materialization='table'` behave only full replace, so it going to be a breaking change
3. Enhance document so that user can understand the behavior based on the combination of options

Note:

- `partition_by` option should be supported on `materialization='table'` (full replace) or `incremental_strategy='merge'` too

**Comment by jtcohen6** [comment]
@tmshn Thanks for the detailed issue!

Ingestion time partitioning (i.e. `materialization='table'` with `partitions` specified) is something we added to dbt a _long_ time ago (v0.10), and we'd actually like to rip it out in its entirety (https://github.com/dbt-labs/dbt/issues/2332). We'd also remove the [corresponding docs](https://docs.getdbt.com/docs/guides/database-specific-guides/creating-date-partitioned-tables), which are currently just unlisted.

In lieu of ingestion-time partitioning, we believe everyone should be using **column-based partitioning**. To enable it, the only dbt configuration you need is [`partition_by`](https://docs.getdbt.com/reference/resource-configs/bigquery-configs#partition-clause)! It's totally up to you whether you materialize the model as `table` or `incremental`, whether you use `merge` or `insert_overwrite` as your incremental strategy, or (in the case of `insert_overwrite`) whether you explicitly specify the partitions to overwrite (by defining `partitions`) or let dbt figure it out dynamically.

So, I wholeheartedly agree with problems 1 + 2; the solution will be to do away with ingestion-partitioned tables. As far as problem 3, could you give more details about the bugs you've observed here?

> 1. it raises type error, 2: it trys to insert data out of specified partition

As for the suggestions:
1. I'm not sure I understand the implications of a user specifying `partition_by = {'field': '_PARTITIONTIME'}`, if it's not actually an ingestion-partitioned table.
2. Agree! We'll accomplish this by removing support for ingestion-partitioned tables.
3. It's clear from problem 4 + suggestion 3 in this issue, and lots of other issues/comments/threads like it, that we need better documentation about all of this! We have reference docs today for [table partitioning](https://docs.getdbt.com/reference/resource-configs/bigquery-configs#using-table-partitioning-and-clustering) and [incremental strategies](https://docs.getdbt.com/reference/resource-configs/bigquery-configs#merge-behavior-incremental-models), but we really need an end-to-end guide that walks users through the matrix of possible configurations, and is something we can keep more up-to-date than [this discourse post](https://docs.getdbt.com/reference/resource-configs/bigquery-configs#using-table-partitioning-and-clustering).

In conclusion:
- Most of your technical proposals are already implemented
- The matrix of configurations is not clearly documented
- There may be a few buggy edge cases with `incremental_strategy: insert_overwrite` + specific values of `partitions` + `partition_by`. If you could open those bug reports as new issues, I'd really appreciate it!

In the meantime, I'm going to transfer this issue to the docs.getdbt.com repo, since that will be the place to contribute improvements to documentation.
